src/protocol/find-tests.ts

```diff
@@ -0,0 +1,445 @@
+    const params: AnalysisParams = {
+      sessionId: this.sessionId,
+      mapper: "",
+      effectful: true,
+      locations: [{ location: this.catchBlockLocation }],
+    };
+
+    const handler: AnalysisHandler<void> = {};
+    const failurePoints: PointDescription[] = [];
+
+    handler.onAnalysisPoints = points => failurePoints.push(...points);
+    await analysisManager.runAnalysis(params, handler);
+
+    return failurePoints;
+  }
+
+  // Get a sorted array of the points where exceptions were thrown.
+  private async getExceptionPoints(): Promise<PointDescription[]> {
+    const params: AnalysisParams = {
+      sessionId: this.sessionId,
+      mapper: "",
+      effectful: true,
+      exceptionPoints: true,
+    };
+
+    const handler: AnalysisHandler<void> = {};
+    const exceptionPoints: PointDescription[] = [];
+
+    handler.onAnalysisPoints = points => exceptionPoints.push(...points);
+    await analysisManager.runAnalysis(params, handler);
+
+    exceptionPoints.sort((a, b) => comparePoints(a.point, b.point));
+
+    return exceptionPoints;
+  }
+
+  private mostRecentTest(point: ExecutionPoint): JestTestInfo | null {
+    for (let i = 0; i < this.tests.length; i++) {
+      if (pointPrecedes(point, this.tests[i].startPoint.point)) {
+        return i ? this.tests[i - 1] : null;
+      }
+    }
+    return null;
+  }
+
+  private mostRecentPoint(
+    point: ExecutionPoint,
+    pointArray: PointDescription[]
+  ): PointDescription | null {
+    for (let i = 0; i < pointArray.length; i++) {
+      if (pointPrecedes(point, pointArray[i].point)) {
+        return i ? pointArray[i - 1] : null;
+      }
+    }
+    return null;
+  }
+}
+
+// Look for a source containing the specified pattern.
+async function findMatchingSourceId(pattern: string): Promise<string | null> {
+  await ThreadFront.ensureAllSources();
+  for (const [sourceId, source] of ThreadFront.sources.entries()) {
+    if (source.url?.includes(pattern)) {
+      return sourceId;
+    }
+  }
+  return null;
+}
+
+// Get the location of the first breakpoint on the specified line.
+// Note: This requires a linear scan of the lines containing breakpoints
+// and will be inefficient if used on large sources.
+async function getBreakpointLocationOnLine(
+  sourceId: string,
+  targetLine: number
+): Promise<Location | null> {
+  const breakpointPositions = await ThreadFront.getBreakpointPositionsCompressed(sourceId);
+  for (const { line, columns } of breakpointPositions) {
+    if (line == targetLine && columns.length) {
+      return { sourceId, line, column: columns[0] };
+    }
+  }
+  return null;
+}
+
+// Look for places in the recording used to run jest tests.
+async function setupJestTests(): Promise<JestTestState | null> {
+  // Look for a source containing the callAsyncCircusFn function which is used to
+  // run tests using recent versions of Jest.
+  const circusUtilsSourceId = await findMatchingSourceId("jest-circus/build/utils.js");
```

Comment:
note that we might start bundling (https://github.com/facebook/jest/pull/12348) at which point at least the filename will break and _possibly_ the function name (depending on sourcemaps).

Not sure how to handle that. It would be better to have a real hook for this (e.g. [`handleTestEvent` in the env](https://jestjs.io/docs/configuration#testenvironment-string)?), but that might not be feasible - I have no idea how this project works.

Reply:
Definitely, the pattern matching on file names and code patterns in this PR is not robust at all, and longer term it would be great to have a way of identifying places where tests started and failed that doesn't rely on detecting these patterns.  We have a mechanism to add annotations to a recording which could be used to find these points.  It isn't currently exposed in node, but if it was we could have something like:

```
if (process.addRecordReplayAnnotations) {
  process.addRecordReplayAnnotation("JestTestStart", { ...info about test... });
}
```

I don't know if it would be possible to use `handleTestEvent` for this --- we would need these annotations to be as close as possible to where the test starts running, so we can find the entry point to the test function which was passed in.  Right now we pattern match to detect the calls to the test function, and then step in to each of those calls to find the point to jump to when someone clicks on that test.

Reply:
We have [`--testLocationInResults`](https://jestjs.io/docs/cli#--testlocationinresults). Would that be enough? I would assume that info is also included in a `handleTestEvent` call, but I haven't verified. In that case you can at the very least point to where the test function is

Reply:
I think that would be enough.  If the test event callback is notified about every test, and has the name of that test, the file/line/column of the test callback, and whether the test passed, that should be enough information to reconstruct everything we're getting from the pattern matching we're doing.  The test callback would call process.addRecordReplayAnnotation with this information, and then the devtools can read that from the recording and use it to figure out where the tests ran.

Reply:
Yeah, its called when test is declared, execution starts, execution enda etc. I'm not 100% sure the location info is present for all cases, but if not we can probably fix that